### ISSUE TITLE
bug(language-service) components told as not included until restart

### ISSUE THREAD

**Issue by cyrilletuzi** [angular/angular#14961]
**I'm submitting a ...**
```
[X] bug report
```

**Current behavior**
I'm using `@angular/language-service` directly from TypeScript 2.2 and 2.3 language service extension, with that line in `tsconfig.json` : 
`"plugins": [{ "name": "@angular/language-service" }]`

When creating a new component and declaring it in some module (manually or with `ng g component test`) `@angular/language-service` raises an error, telling : 

`[ts] Component 'TestComponent' is not included in a module and will not be available inside a template.`

But it is. The app works very well in the browser. The only way to get rid of the error is to restart the editor (VS Code in my case).

I use the language service in my courses, to help the students, and it's a blocking issue, as it confuses them instead of helping them. And it's blocking for everyone, as we can't seriously restart our editor each time we create a component.

Same for services and others.

Another consequence of this : the new component is not suggested by Intellisense in other templates.

**Expected behavior**
Declared components (and others) should not raise an error. `@angular/language-service` should keep its internal check up to date in real time to know what is declared or not.

And Intellisense should be up to date in real time too.

* **Angular version:** 4.0.0-rc.2
 
* **Language:** TypeScript 2.2.1 and 2.3.0-dev.20170306

**Comment by cyrilletuzi** [comment]
After investigation, there is a `updateAnalyzedModules()` method, but only called in tests. Not sure where it should be called so the declarations diagnostics are always up to date.

**Comment by chuckjaz** [comment]
@cyrilletuzi The intent is for the user of the language service `updateAnalyzedModules()` during a quite period. This method can take over 800ms to complete so it shouldn't be called too often (after 2-3 seconds or so of no other activity). 

I will be working on improving this in for 4.1. It shouldn't take as long as it does to calculate the modules and certainly we can cache the information from the previous call. Once these changes are done `updateAnalyzedModules()` will be much faster and it will no longer required, but still recommended.

**Comment by cyrilletuzi** [comment]
`updateAnalyzedModules()` by the user is OK with a editor plugin. But the VS Code plugin is still not stable (and won't be soon, if I've understood the plans), and it was a temporary workaround for native TypeScript support.

Now that we can do `"plugins": [{ "name": "@angular/language-service" }]` in `tsconfig.json`, I have no way to trigger `updateAnalyzedModules()` by myself.

Of course, if it's time expansive, it should not trigger every second, but at least when necessary (some delay ? when change of file ? when new file ?).

For me it should not wait for 4.1 and be ready for 4.0, as @IgorMinar set it first.

I do Angular 2 courses from 1 year now, and missing autocomplete in templates is **always** a reproach done by programers. So last time I tried the language service, but as it triggers fatal errors when there is none, it was confusing them or even blocking even instead of helping them.

(Previous time, I tried to use the VS Code plugin, but it has serious performance issues, on medium hardware computers it was so heavy all started to freeze, so we had to remove it.)

So next course I'll be forced to not use the language service, as it's unusable in it's current form. I think it's too bad we have this module since 2.4, and that it still won't be usable in 4.0.

**Comment by chuckjaz** [comment]
Sorry. I misunderstood how you are using this.

We will be fixing this for 4.0.1.

I am planning on doing significant upgrades to the `vscode` plugin after 4.0.0 lands to fix a number of the outstanding issues including the freezing.

The work that is slated for 4.1 is improving the API that editor integrators use to incorporate the language service into new editors (which I mistakenly thought you were doing).

**Comment by cyrilletuzi** [comment]
Is #15543 (released in 4.0.1) supposed to fix this issue ? Because it's still there.

Based on the PR info, I tried to manually modify the component file and/or the module file, but the component error still stays until I reload my editor.

**Comment by chuckjaz** [comment]
No. This bug is still open. I intend to fix this this week.

**Comment by cyrilletuzi** [comment]
A little worried to not see this issue closed for 4.1-rc.0, is it still on track ? Otherwise it means it will be another Angular version where the language service is not really usable natively with TS plugins. (And the VS Code extension is not yet usable either, it crashes quite every time.)

**Comment by chuckjaz** [comment]
This requires a change to the TypeScript language service that I am currently working with the TypeScript team to land. Once the change in TypeScript lands then the fix can land here.

**Comment by cyrilletuzi** [comment]
Thanks. Is it planned for TS 2.3 ?

**Comment by chuckjaz** [comment]
I believe so. I have a PR out now.

**Comment by bhagatparwinder** [comment]
Facing the exact issue as mentioned by @cyrilletuzi with TypeScript 2.3.3 and Angular 4.1.3
For me, it does not go away on restart.

**Comment by chuckjaz** [comment]
@bhagatparwinder If it doesn't go away on restart it is a different issue. Can you report it as a separate issue with steps to reproduce?

**Comment by sandangel** [comment]
I have the same issue here. I'm using typescript 2.4.2 and angular 4.4rc-0. In the output tab, I notice that it has a punch of output seem like a crash and this bug happened. then when I restart vscode, it back to normal. this happen after I had edit my code for 2 or 3 hour so I don't know how to reproduce this issue. Sorry about that.

**Comment by bhagatparwinder** [comment]
@chuckjaz It is working fine now 👍

**Comment by ivanlopezmolina** [comment]
I was having the same issue 

Component 'TestComponent' is not included in a module and will not be available inside a template.

and while I was researching an reading some threads of pretty much the same issue, the error suddenly (after about 15 minutes) disappeared, I think is more a timing issue when you change/add components

**Comment by cyrilletuzi** [comment]
@bhagatparwinder with which version of Angular and TypeScript ?

**Comment by bhagatparwinder** [comment]
@cyrilletuzi I didn't update either. I'm still on TypeScript 2.3.3 and Angular 4.1.3 with Language Service 0.1.7

**Comment by cyrilletuzi** [comment]
@chuckjaz Is there anything new about this issue ?

Even with the VS Code plugin, the detection of the declaration of components is not reactive at all.

It's quite annoying for new Angular users, who think there is an error when there is none.

**Comment by kyliau** [comment]
Looks like this is related to https://github.com/angular/angular/issues/19405, for which I have a pending PR. Will discuss with Chuck and see if the fix will negatively impact the performance

**Comment by kyliau** [comment]
@cyrilletuzi commit https://github.com/angular/angular/commit/cc9419d1cababe1bbf366f982517c6290e7aeff8 should have fixed this issue. Please let me know otherwise.

**Comment by cyrilletuzi** [comment]
@kyliau Tested with Angular 5.2, TypeScript 2.5.3, and language service as a TypeScript plugin in `tsconfig.json`, and the problem is still there. :(

**Comment by mlc-mlapis** [comment]
@cyrilletuzi ... but the base of the problem is in the VSC extension ... that actually is not able to use the actual `@angular/language-service` lib.

**Comment by cyrilletuzi** [comment]
@mlc-mlapis No, I tried without the VSC extension, using the language service as a TypeScript plugin in tsconfig.json :

```
{
  "compilerOptions": {
    "plugins": {
       "name": "@angular/language-service"
    }
  }
} 
```

**Comment by mlc-mlapis** [comment]
@cyrilletuzi ... ah, ok then.

**Comment by cyrilletuzi** [comment]
#21337 fixed this in 5.2.1. Finally! Thanks!

For everyone : it now means you can use the language service directly as a TS plugin, without a VS Code extension, which should be much faster !

But an important note : you need to tell VS Code to use the TS of your project (meaning currently TS 2.5) and not the one of VS Code (currently 2.6), because Angular and the language service are not yet compatible with TS 2.6.

**Comment by jtsom** [comment]
Curious as to how to get this working.. I removed the VS Code extension, and added the plugins item in the tsconfig.json, but I don't see any hovers / intellisense. I made sure my VS Code is using typescript 2.5.3

**Comment by cyrilletuzi** [comment]
I suppose you use external templates. The VS Code extension is still needed for them, as it required VS Code specific code, plugins in TS config only work in `.ts` files, so for inline templates.

**Comment by jtsom** [comment]
Ah, yeah, it's an Angular app, so the templates are not in the same file. Got it, thanks.

**Comment by Matmo10** [comment]
Everything mentioned so far seems to be VSCode specific. How do we fix this for IntelliJ/Webstorm?

**Comment by linocatucci** [comment]
Hello, 
the problems is still there with WebStorm 2017.3.3
Build #WS-173.4301.22, built on January 15, 2018
Subscription is active until September 16, 2018
JRE: 1.8.0_152-release-1024-b11 x86_64
JVM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
Mac OS X 10.13.2

**Comment by taitgremp** [comment]
Still present

**Comment by jarcko** [comment]
I have the same issue,

WebStorm 2018.1.5
Build #WS-181.5281.31, built on June 14, 2018
Licensed to Ciklum TOB / Igor Jarko
You have a perpetual fallback license for this version
Subscription is active until March 31, 2019
JRE: 1.8.0_152-release-1136-b39 amd64
JVM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
Windows 8.1 6.3

**Comment by chico1198** [comment]
I also have the same issue:

IntelliJ IDEA 2018.2.3 (Ultimate Edition)
Build #IU-182.4323.46, built on September 3, 2018
JRE: 1.8.0_152-release-1248-b8 x86_64
JVM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
Mac OS X 10.11.6
